File: 2023/05/go/main.go
package main

import (
	core "advent/of/go/src/go"
	"errors"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

type rRange struct {
	Source      rMinMax
	Destination rMinMax
}

type rMinMax struct {
	Min int64
	Max int64
}

func parseInput(input string, single bool) ([]rMinMax, [][]rRange) {
	parsed := strings.Split(input, "\n\n")
	var header []int64
	for _, v := range strings.Split(strings.Split(parsed[0], ":")[1], " ") {
		if len(v) > 0 {
			val, _ := strconv.ParseInt(v, 10, 64)
			header = append(header, val)
		}
	}
	var seedRanges []rMinMax
	if single {
		for _, v := range header {
			seedRanges = append(seedRanges, rMinMax{Min: v, Max: v})
		}
	} else {
		for i := 0; i < len(header); i += 2 {
			seedRanges = append(seedRanges, rMinMax{Min: header[i], Max: header[i] + header[i+1]})
		}
	}

	var srcToDestRanges [][]rRange
	for _, v := range parsed[1:] {
		var tr []rRange
		for _, w := range strings.Split(v, "\n")[1:] {
			var r rRange
			x := strings.Split(w, " ")

			dest, _ := strconv.ParseInt(x[0], 10, 64)
			src, _ := strconv.ParseInt(x[1], 10, 64)
			inc, _ := strconv.ParseInt(x[2], 10, 64)

			r.Source = rMinMax{Min: src, Max: src + inc - 1}
			r.Destination = rMinMax{Min: dest, Max: dest + inc - 1}

			tr = append(tr, r)
		}
		srcToDestRanges = append(srcToDestRanges, tr)
	}

	return seedRanges, srcToDestRanges
}

func find(groups []rRange, r rMinMax) (rRange, error) {
	for _, g := range groups {
		if g.Source.Min <= r.Min &&
			r.Min <= g.Source.Max &&
			r.Max >= g.Source.Min &&
			g.Source.Max >= r.Max {
			return g, nil
		}
	}
	return rRange{}, errors.New("not found")
}

func solve(input string, single bool) string {
	seedRanges, srcToDestRanges := parseInput(input, single)

	for _, groups := range srcToDestRanges {
		for _, g := range groups {
			newSeeds := []rMinMax{}
			for i := 0; i < len(seedRanges); i++ {
				r := &seedRanges[i]
				if r.Min < g.Source.Min && g.Source.Min < r.Max {
					newSeeds = append(newSeeds, rMinMax{Min: r.Min, Max: g.Source.Min - 1})
					r.Min = g.Source.Min
				}
				if r.Min < g.Source.Max && g.Source.Max < r.Max {
					newSeeds = append(newSeeds, rMinMax{Min: g.Source.Max + 1, Max: r.Max})
					r.Max = g.Source.Max
				}
			}
			seedRanges = append(newSeeds, seedRanges...)
		}
		for i := 0; i < len(seedRanges); i++ {
			r := &seedRanges[i]
			found, err := find(groups, *r)
			if err != nil {
				continue
			}
			diff := found.Destination.Min - found.Source.Min
			r.Min += diff
			r.Max += diff
		}
	}

	var res int64 = math.MaxInt64
	for _, r := range seedRanges {
		if res > r.Min {
			res = r.Min
		}
	}
	return strconv.FormatInt(res, 10)
}

func part1(input string, isTest bool) string {
	return solve(input, true)
}

func part2(input string, isTest bool) string {
	return solve(input, false)
}

File: 2023/05/lua/main.lua
local run = require('src.lua.run')
local str = require('src.lua.utils.str')

local options = {
   has_alternate = false,
   has_io = false,
}

local function parse_input(input, single)
   local parsed = {}
   input = input:gsub('\n\n', '|')
   for line in input:gmatch('[^|]+') do
      parsed[#parsed + 1] = line
   end

   local header = {}
   for i, v in ipairs(str.split(str.split(parsed[1], ':')[2], ' ')) do
      if #v > 0 then
         header[#header + 1] = tonumber(v)
      end
   end

   local seed_ranges = {}
   if single then
      for _, v in ipairs(header) do
         seed_ranges[#seed_ranges + 1] = { min = v, max = v }
      end
   else
      for i = 1, #header, 2 do
         seed_ranges[#seed_ranges + 1] = { min = header[i], max = header[i] + header[i + 1] }
      end
   end

   local src_to_dest_ranges = {}
   for i = 2, #parsed do
      local tr = {}
      local v = parsed[i]
      local w = str.split(v, '\n')
      for j = 2, #w do
         local z = str.split(w[j], ' ')
         local r = {}

         local dest = tonumber(z[1])
         local src = tonumber(z[2])
         local inc = tonumber(z[3])

         r.source = { min = src, max = src + inc - 1 }
         r.destination = { min = dest, max = dest + inc - 1 }

         tr[#tr + 1] = r
      end

      src_to_dest_ranges[#src_to_dest_ranges + 1] = tr
   end

   return seed_ranges, src_to_dest_ranges
end

local function find_range(groups, r)
   for _, g in ipairs(groups) do
      if g.source.min <= r.min and
          r.min <= g.source.max and
          r.max >= g.source.min and
          g.source.max >= r.max then
         return g
      end
   end
end

local function solve(input, single)
   local seed_ranges, src_to_dest_ranges = parse_input(input, single)

   for _, groups in ipairs(src_to_dest_ranges) do
      for _, g in ipairs(groups) do
         for i = 1, #seed_ranges do
            local r = seed_ranges[i]
            if r.min < g.source.min and g.source.min < r.max then
               seed_ranges[#seed_ranges + 1] = { min = r.min, max = g.source.min - 1 }
               r.min = g.source.min
            end
            if r.min < g.source.max and g.source.max < r.max then
               seed_ranges[#seed_ranges + 1] = { min = g.source.max + 1, max = r.max }
               r.max = g.source.max
            end
         end
      end
      for i = 1, #seed_ranges do
         local r = seed_ranges[i]
         local found = find_range(groups, r)
         if found == nil then
            goto continue
         end
         local diff = found.destination.min - found.source.min
         r.min = r.min + diff
         r.max = r.max + diff
         ::continue::
      end
   end

   local res = math.huge

   for _, r in ipairs(seed_ranges) do
      if res > r.min then
         res = r.min
      end
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   return solve(input, true)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   return solve(input, false)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()
